src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -1137,10 +1137,27 @@ public static unsafe SocketError SendAsync(SafeSocketHandle handle, IList<ArrayS
                 transmitFileBuffers.TailLength = postBuffer.Length;
             }
 
-            bool success = Interop.Mswsock.TransmitFile(socket, fileHandle, 0, 0, overlapped,
-                needTransmitFileBuffers ? &transmitFileBuffers : null, flags);
+            bool releaseRef = false;
+            IntPtr fileHandlePtr = IntPtr.Zero;
+            try
+            {
+                if (fileHandle != null)
+                {
+                    fileHandle.DangerousAddRef(ref releaseRef);
+                    fileHandlePtr = fileHandle.DangerousGetHandle();
```

Comment:
Maybe a dumb question, but is there no better way to deal with a possibly null handle than to do it manually? Seems like we should be able to pass null or some special SafeHandle value here...

Reply:
Not that I'm aware of.  It's pretty explicit:
https://github.com/dotnet/runtime/blob/e489111db3cb25a7ff9640c2e3a6e9561f56cc07/src/coreclr/src/System.Private.CoreLib/src/System/StubHelpers.cs#L1251-L1256
cc: @AaronRobinsonMSFT, @jkoritzinsky

Reply:
I can understand why null is not supported, but it seems like it would be straightforward to define something like SafeHandle.ZeroHandle or whatever that always just passes a zero handle.

Reply:
@geoffkizer @stephentoub Is the thought here we should let the caller pass `null`? I'm not technically against this but it would be a chance in behavior. I could see an argument that this code path should be "as fast as possible" and could see assume the best case always.

@jkoritzinsky Your thoughts? Perhaps we can address this in the source gen approach for .NET 6?

/cc @elinor-fung

Reply:
I just want a way to pass a zero handle without having to remove the SafeHandle from the p/invoke call and do the addref/release manually, because this is fragile.

Reply:
> it seems like it would be straightforward to define something like SafeHandle.ZeroHandle or whatever that always just passes a zero handle

In most P/Invokes I've seen, as is the case here, the DllImport is strongly-typed to the SafeHandle-derived type, so you'd need to have such a ZeroHandle on each one.  (Lots of SafeHandles also consider 0 to be an invalid handle value, though I don't think that'd be an issue for this particular use case.)

For this particular case, we could add an internal ZeroHandle to SafeSocketHandle, but I'm not sure that really buys us much, given it's used once.

> Is the thought here we should let the caller pass null?

Yeah, I think that'd be useful.  This isn't the first time I've seen this, though generally I've seen it manifest as someone defining two different DllImport overloads, one that takes IntPtr and one that takes SafeHandle, to work around this.  In fact that's what the equivalent code in .NET Framework does here (which we regressed at some point, leading to this fix).

Reply:
@geoffkizer or @stephentoub  Feel free to file an issue for this. I think it warrants a serious discussion because I agree this is annoying.

Reply:
Opened https://github.com/dotnet/runtime/issues/42550. Thanks.